src/main/java/nwoolcan/model/brewery/warehouse/article/ArticleManager.java

```diff
@@ -0,0 +1,137 @@
+        if (!articleToId.containsKey(article)) {
+            int newId = nextAvailableId++;
+            article = new ArticleImpl(newId, name, unitOfMeasure);
+            articleToId.put(article, newId);
+            idToArticle.put(newId, article);
+        }
+        return idToArticle.get(articleToId.get(article));
+    }
+    /**
+     * Creates a {@link BeerArticle} and insert it into the {@link ArticleManager}. If it already exists, it will be returned.
+     * @param name of the {@link BeerArticle}.
+     * @param unitOfMeasure the {@link UnitOfMeasure} of the {@link BeerArticle}.
+     * @return a new {@link BeerArticle} if it does not exist in the {@link ArticleManager}, otherwise the existing one.
+     */
+    @SuppressWarnings("NullAway")
+    public synchronized BeerArticle createBeerArticle(final String name,
+                                                      final UnitOfMeasure unitOfMeasure) {
+        Article beerArticle = new BeerArticleImpl(fakeId, name, unitOfMeasure);
+        if (!articleToId.containsKey(beerArticle)) {
+            int newId = nextAvailableId++;
+            beerArticle = new BeerArticleImpl(newId, name, unitOfMeasure);
+            articleToId.put(beerArticle, newId);
+            idToArticle.put(newId, beerArticle);
+        }
+        return (BeerArticle) idToArticle.get(articleToId.get(beerArticle));
+    }
+    /**
+     * Creates a {@link IngredientArticle} and insert it into the {@link ArticleManager}. If it already exists, it will be returned.
+     * @param name of the {@link IngredientArticle}.
+     * @param unitOfMeasure the {@link UnitOfMeasure} of the {@link IngredientArticle}.
+     * @param ingredientType the {@link IngredientType} of the {@link IngredientArticle}.
+     * @return a new {@link IngredientArticle} if it does not exist in the {@link ArticleManager}, otherwise the existing one.
+     */
+    @SuppressWarnings("NullAway")
+    public synchronized IngredientArticle createIngredientArticle(final String name,
+                                                                  final UnitOfMeasure unitOfMeasure,
+                                                                  final IngredientType ingredientType) {
+        Article ingredientArticle = new IngredientArticleImpl(fakeId, name, unitOfMeasure, ingredientType);
+        if (!articleToId.containsKey(ingredientArticle)) {
+            int newId = nextAvailableId++;
+            ingredientArticle = new IngredientArticleImpl(newId, name, unitOfMeasure, ingredientType);
+            articleToId.put(ingredientArticle, newId);
+            idToArticle.put(newId, ingredientArticle);
+        }
+        return (IngredientArticle) idToArticle.get(articleToId.get(ingredientArticle));
+    }
+    /**
+     * Return a {@link Set} of all the managed {@link Article}.
+     * @return a {@link Set} of all the managed {@link Article}.
+     */
+    public Set<Article> getArticles() {
```

Comment:
I think articles should be filtered in the articles view (for name, type, etc...).
I consider using the Query Pattern if needed (hence, a `QueryArticle` class).

Reply:
I left the filtering/ordering part to the warehouse as it was before, with a QueryArticle class. I don't know if it is good to assign that responsibility to the article manager. Presuming you didn't see it in warehouse, what do you think about that?

Reply:
Yep didn't see